### PR TITLE
Track active proposal balance for governor

### PR DESCRIPTION
**Pull request by amandeeprehal-lab49**
Track active proposal balance for governor

### CODE REVIEW

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 1]
```diff
@@ -435,16 +447,24 @@ abstract contract GovernorCountingSimpleInternal is
...
         info = proposals[proposalId];
-        require(
-            info.creator != address(0),
-            "GovernorCountingSimpleInternal: Proposal not found"
-        );
+        require(info.creator != address(0), "GCSI: Proposal not found");
+    }
+
+    function getBlockedTokenBalance() internal view returns (uint256) {
+        return activeProposalsCount * PROPOSAL_CREATION_AMOUNT;
```
Comment (shivamgarglab49) [inline contracts/governance/GovernorCountingSimpleInternal.sol:467]: I think we should return the amount based on token type :
1- for FT  = activeProposalsCount * PROPOSAL_CREATION_AMOUNT
2- for NFT = activeProposalsCount only

File: test/governor-test.ts [hunk 2]
```diff
@@ -148,6 +148,15 @@ describe("Governor Tests", function () {
     };
   }
 
+  const readLastGovernorBalanceEvent = async (txn: any) => {
+    const events = await TestHelper.readEvents(txn, "GovernorBalance");
```
Comment (shivamgarglab49) [inline test/governor-test.ts:152]: please pass "GovernorBalance" in array ["GovernorBalance"]

File: contracts/governance/GovernorTransferToken.sol [hunk 3]
```diff
@@ -109,10 +113,13 @@ contract GovernorTransferToken is GovernorCountingSimpleInternal {
...
 
     function _transfer(bytes memory _data) private {
-        (, address to, address token, uint256 amount) = abi.decode(
-            _data,
-            (uint256, address, address, uint256)
-        );
+        (, address to, address token, uint256 amount, int32 _tokenType) = abi
+            .decode(_data, (uint256, address, address, uint256, int32));
+
+        if (_tokenType == 0) {
```
Comment (shivamgarglab49) [inline contracts/governance/GovernorTransferToken.sol:119]: what about NFT, this checking only for FT ?

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 4]
```diff
@@ -66,12 +66,17 @@ abstract contract GovernorCountingSimpleInternal is
         uint256 amountOrId
     );
 
+    event GovernorBalance(
```
[the remark this answers is not in the document]
Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorCountingSimpleInternal.sol:69]: Done

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 5]
```diff
@@ -170,6 +175,13 @@ abstract contract GovernorCountingSimpleInternal is
             getVotingInformation(proposalId),
             amountOrId
         );
+
+        activeProposalsCount = activeProposalsCount + 1;
```
[the remark this answers is not in the document]
Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorCountingSimpleInternal.sol]: Done

File: contracts/governance/GovernorCountingSimpleInternal.sol (hunk 1 above)
Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorCountingSimpleInternal.sol]: Done

File: test/governor-test.ts (hunk 2 above)
Comment (amandeeprehal-lab49) [reply test/governor-test.ts]: Done

File: contracts/governance/GovernorTransferToken.sol (hunk 3 above)
Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorTransferToken.sol]: Done

File: contracts/governance/GovernorCountingSimpleInternal.sol [hunk 6]
```diff
@@ -434,16 +437,28 @@ abstract contract GovernorCountingSimpleInternal is
...
         info = proposals[proposalId];
-        require(
-            info.creator != address(0),
-            "GovernorCountingSimpleInternal: Proposal not found"
-        );
+        require(info.creator != address(0), "GCSI: Proposal not found");
+    }
+
+    function getBlockedTokenBalance() internal returns (uint256) {
+        int32 tokenType = _tokenType(hederaService, token);
```
Comment (shivamgarglab49) [inline contracts/governance/GovernorCountingSimpleInternal.sol:457]: we hitting the token API every time here.

May be we can use _tokenHolder to check same like we doing for other places.

Comment (amandeeprehal-lab49) [reply contracts/governance/GovernorCountingSimpleInternal.sol:457]: Done

**Review by shivamgarglab49 — APPROVED** [review]

**Pull request merged**